reconstructSPI/iterative_refinement/expectation_maximization.py

```diff
@@ -504,10 +505,14 @@ def insert_slice(slice_real, xyz, n_pix):
             otherwise 0.
             Shape (n_pix, n_pix, n_pix)
         """
-        shape = xyz.shape[0]
-        count_3d = np.ones((n_pix, n_pix, n_pix))
-        count_3d[0, 0, 0] *= shape
-        inserted_slice_3d = np.ones((n_pix, n_pix, n_pix))
+        inserted_slice_tensor = torch.sparse_coo_tensor(
+            xyz + n_pix // 2, slice_real.reshape((n_pix**2,)), (n_pix, n_pix, n_pix)
+        )
+        count_tensor = torch.sparse_coo_tensor(
+            xyz + n_pix // 2, np.ones((n_pix**2,)), (n_pix, n_pix, n_pix)
+        )
+        inserted_slice_3d = inserted_slice_tensor.to_dense().numpy()
+        count_3d = count_tensor.to_dense().numpy()
```

Comment:
sparse_coo_tensor does not do any interpolation. 

you'll still have to use map_coordinates, which returns a dense 3d array.

until there is a (custom in pytorch?) linear interpolation that returns sparse idx at the inserted voxels, and the value inserted at these voxels. there will be no use for sparse_coo_tensor.

So for now just stick with what I did in the notebook using map_coordinates.

Reply:
Sounds good, problem for tomorrow.

Reply:
@geoffwoollard I've started looking into this and I'd appreciate if you could clarify some points.

I'm not completely certain why we can't use the torch functionality, and where interpolation is needed. So from what I've seen so far, map_coordinates only goes the other direction and doesn't seem to function for putting slices *back* into an array - browsing through scipy I did find https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interpn.html#scipy.interpolate.interpn which can serve the correct purpose (mapping a value onto a given point in a 3D grid). But upon thinking about this, I don't think the interpolation would actually *do* anything here as we have a precise voxel we are pointing to in the xyz array to which we are mapping a value from slice - no interpolation needed so far as I can tell. 

As for torch, what it does right now is precisely that to begin with - mapping a value from the slice to a voxel pointed to by the xyz array. Then, I use the to_dense method to collapse this into a dense 3D array (in the exact same shape and data representation format as the map_3d arrays we are using). 

I guess I fail to see where the need for interpolation comes in.

Reply:
@geoffwoollard I've thought some more on this and I think the problem needs some clarification. 

Unfortunately, map_coordinates won't serve our purposes exactly as it goes the wrong direction (it can only pull points out of an existing grid, not put them back in). The method that I've found that is most tailored to our purposes is scipy.interpolate.griddata. However, when using this method it returned a rather illuminating error - a 3D grid of points cannot be interpolated from just a 2D slice.

Thinking more about this, of course that is true - the problem of interpolating a point in 3D near a plane is ill-defined if that plane has zero width - is the result always zero unless the point is *on* the plane? How can a point be *on* a plane if it has zero width - float precision limitations would mean that all points not precisely defined during the creation of the plane would evaluate to 0 when interpolated.

This is to say that we need to clarify a width for slices if we want to interpolate values from them. Perhaps something dependent on n_pix, say n_pix / 8 or something. What also needs to happen is that we need to generate the coordinates of the other "side" of the slice in order to give it its width - ideally by "tiling" the slice once along its normal.

Reply:
@geoffwoollard What I've done is added a "depth" to the outputted rotated xy plane in generate_slices (the depth is added before rotating). This way, when we go to insert a slice, we have a volume which the slice occupies rather than just a 2D area. since this volume has depth << pixel size by design, it shouldn't affect the result validity. The slice (upon insertion) is assumed to be constant along this depth axis. 

With this change, I was able to get interpolation working for insert_slice.

Reply:
@thisTyler 

Yes, good point about `map_coordinates`. I had used another solution here:
https://github.com/geoffwoollard/learn_cryoem_math/blob/master/src/interp.py
https://github.com/geoffwoollard/learn_cryoem_math/blob/master/nb/fourier_slice_2D_3D_with_trilinear.ipynb

Can you push a notebook or code snippet to this branch illustrating the griddata issue and error. 

It needs to insert density in nearby voxels, so it should be aware of the 3D nature of arr_3d it is inserting into and the "size" between voxels.

Reply:
@geoffwoollard https://github.com/compSPI/reconstructSPI/blob/itr_ref_insert_slice/Insert%20Slice%20Tolerance%20Demo.ipynb